src/wrapper/objects/jvalue.rs

```diff
@@ -185,58 +186,157 @@ impl<'a> From<jboolean> for JValue<'a> {
     fn from(other: jchar) -> Self {
         JValue::Char(other)
     }
 }
 
+impl<'a> TryFrom<JValue<'a>> for jchar {
+    type Error = Error;
+
+    fn try_from(value: JValue<'a>) -> Result<Self> {
+        match value {
+            JValue::Char(c) => Ok(c),
+            _ => Err(Error::WrongJValueType("char", value.type_name())),
+        }
+    }
+}
+
 // jshort
 impl<'a> From<jshort> for JValue<'a> {
     fn from(other: jshort) -> Self {
         JValue::Short(other)
     }
 }
 
+impl<'a> TryFrom<JValue<'a>> for jshort {
+    type Error = Error;
+
+    fn try_from(value: JValue<'a>) -> Result<Self> {
+        match value {
+            JValue::Short(s) => Ok(s),
+            _ => Err(Error::WrongJValueType("short", value.type_name())),
+        }
+    }
+}
+
 // jfloat
 impl<'a> From<jfloat> for JValue<'a> {
     fn from(other: jfloat) -> Self {
         JValue::Float(other)
     }
 }
 
+impl<'a> TryFrom<JValue<'a>> for jfloat {
+    type Error = Error;
+
+    fn try_from(value: JValue<'a>) -> Result<Self> {
+        match value {
+            JValue::Float(f) => Ok(f),
+            _ => Err(Error::WrongJValueType("float", value.type_name())),
+        }
+    }
+}
+
 // jdouble
 impl<'a> From<jdouble> for JValue<'a> {
     fn from(other: jdouble) -> Self {
         JValue::Double(other)
     }
 }
 
+impl<'a> TryFrom<JValue<'a>> for jdouble {
+    type Error = Error;
+
+    fn try_from(value: JValue<'a>) -> Result<Self> {
+        match value {
+            JValue::Double(d) => Ok(d),
+            _ => Err(Error::WrongJValueType("double", value.type_name())),
+        }
+    }
+}
+
 // jint
 impl<'a> From<jint> for JValue<'a> {
     fn from(other: jint) -> Self {
         JValue::Int(other)
     }
 }
 
+impl<'a> TryFrom<JValue<'a>> for jint {
+    type Error = Error;
+
+    fn try_from(value: JValue<'a>) -> Result<Self> {
+        match value {
+            JValue::Int(i) => Ok(i),
+            _ => Err(Error::WrongJValueType("int", value.type_name())),
+        }
+    }
+}
+
 // jlong
 impl<'a> From<jlong> for JValue<'a> {
     fn from(other: jlong) -> Self {
         JValue::Long(other)
     }
 }
 
+impl<'a> TryFrom<JValue<'a>> for jlong {
+    type Error = Error;
+
+    fn try_from(value: JValue<'a>) -> Result<Self> {
+        match value {
+            JValue::Long(l) => Ok(l),
+            _ => Err(Error::WrongJValueType("long", value.type_name())),
+        }
+    }
+}
+
 // jbyte
 impl<'a> From<jbyte> for JValue<'a> {
     fn from(other: jbyte) -> Self {
         JValue::Byte(other)
     }
 }
 
+impl<'a> TryFrom<JValue<'a>> for jbyte {
+    type Error = Error;
+
+    fn try_from(value: JValue<'a>) -> Result<Self> {
+        match value {
+            JValue::Byte(b) => Ok(b),
+            _ => Err(Error::WrongJValueType("byte", value.type_name())),
+        }
+    }
+}
+
 // jvoid
 impl<'a> From<()> for JValue<'a> {
     fn from(_: ()) -> Self {
         JValue::Void
     }
 }
+
+impl<'a> TryFrom<JValue<'a>> for () {
```

Comment:
Is `TryFrom` implementation needed for `()`? Are there any cases when a user _expects_ Void and needs to do try_from to get void — essentially, nothing?

Reply:
It's not really needed, it's just for... symmetry?
The only way I think it could be useful is to check for Java exceptions when calling a method in a generic way, even for void-returning methods, but I understand it may not be that useful.

Reply:
To be more concrete: I could easily write something like (not actual code)
```rust
struct AJavaGenericFunction; // closure-like struct but for a static function, so no fields

impl JavaCall {
    fn<T, U, 'a> call_generic_fn(&self, arg: T) -> Result<U> where T: Into<JValue<'a>>,  U: Into<JValue<'a>> + SomeOtherTrait {
        let env = self.get_env();
        env.call_static_method_unchecked(self.into(), self.get_fn_method_id(), U::get_java_type(), &[arg.into()]).map(TryFrom::try_from)
    }
}

// some lines below
let arg: jint = 3;
let v: jint = java_call.call_generic_fn(arg)?;
```


Which would make managing types much more hassle free by having a universal type-conversion utility function.

Actually it's [something I'm looking into automating](https://github.com/giovanniberti/robusta) via a derive macro (not in the form outlined above, it's just to offer some plausible use-case for the `Void` impl)

Reply:
I see, thanks for the explanation. I do not object to adding it to enable more concise client-side code.